### PR TITLE
IGNITE-24076 Distributed properties to block new connections

### PR DESCRIPTION
**Pull request by nizhikov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerConnectionContext.java [hunk 1]
```diff
@@ -88,4 +90,13 @@ void initializeFromHandshake(GridNioSession ses, ClientListenerProtocolVersion v
      * Connection attributes.
      */
     Map<String, String> attributes();
+
+    /**
+     * @return {@code True} if client is management.
+     */
+    default boolean isManagementClient() {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerConnectionContext.java:97]: managementClient

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerConnectionContext.java [hunk 2]
```diff
@@ -88,4 +90,13 @@ void initializeFromHandshake(GridNioSession ses, ClientListenerProtocolVersion v
...
+        Map<String, String> attrs = attributes();
+
+        return attrs != null && Boolean.parseBoolean(attrs.get(MANAGEMENT_CLIENT_ATTR));
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerConnectionContext.java:100]: `attributes()` doesn't return null.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerNioListener.java [hunk 3]
```diff
@@ -534,4 +551,40 @@ private void ensureClientPermissions(byte clientType) throws IgniteCheckedExcept
...
+        boolean isControlUtility = connCtx.clientType() == THIN_CLIENT && connCtx.isManagementClient();
+
+        if (nodeInRecoveryMode()) {
+            if (!isControlUtility)
+                throw new ClientConnectionNodeRecoveryException("Node in recovery mode.");
+
+            return;
+        }
+
+        // If security enabled then only admin allowed to connect as management
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerNioListener.java:571]: dot in the end of line

File: modules/core/src/main/java/org/apache/ignite/internal/cluster/DistributedConfigurationUtils.java [hunk 4]
```diff
@@ -95,4 +106,42 @@ public static <T extends Serializable> IgniteInternalFuture<Void> setDefaultValu
...
+
+        return props;
+    }
+
+    /**
+     * @param prop Property to get value from.
+     * @return Property value. Default {@code true}.
+     */
+    public static boolean asBoolean(@Nullable DistributedBooleanProperty prop) {
+        return prop != null ? prop.getOrDefault(true) : Boolean.TRUE;
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cluster/DistributedConfigurationUtils.java:145]: `prop` never null

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 5]
```diff
@@ -297,6 +301,12 @@ class ServerImpl extends TcpDiscoveryImpl {
     private final ConcurrentMap<InetSocketAddress, GridPingFutureAdapter<IgniteBiTuple<UUID, Boolean>>> pingMap =
         new ConcurrentHashMap<>();
 
+    /** Client node connection allowed property. */
+    private DistributedBooleanProperty clientConnectionEnabled;
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java:305]: Here and below
1. final
2. follow abbreviation rules

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 6]
```diff
@@ -4405,9 +4425,10 @@ else if (log.isDebugEnabled())
                     }
                 }
 
-                IgniteNodeValidationResult err;
+                IgniteNodeValidationResult err = ensureJoinEnabled(node);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java:4428]: Why do you decide to add the check here? There is an alternative - similar to checking node authentication. 

As I can see "validate" is two-factor process - `validateNode(node)`, `validateNode(node, data)`. The logic is described in `GridComponent#validateNode`. Looks it's a not obvious place for this check.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/reader/StandaloneGridKernalContext.java [hunk 7]
```diff
@@ -677,7 +677,7 @@ private void setField(IgniteEx kernal, String name, Object val) throws NoSuchFie
 
     /** {@inheritDoc} */
     @Override public GridInternalSubscriptionProcessor internalSubscriptionProcessor() {
-        return null;
+        return new GridInternalSubscriptionProcessor(this);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/reader/StandaloneGridKernalContext.java:680]: Does standalone instance support any connections? I don't see neither ClientProcessor, nor ability to work with other nodes.

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/reader/StandaloneGridKernalContext.java (hunk 7 above)
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/reader/StandaloneGridKernalContext.java:680]: It's "for test only" change. I will move it to the tests.

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**